src/System.Private.CoreLib/src/System/ValueType.cs

```diff
@@ -36,4 +42,161 @@ public override int GetHashCode()
             return RuntimeAugments.Callbacks.ValueTypeGetHashCodeUsingReflection(this);
         }
     }
+#else
+        private const int UseFastHelper = -1;
+        private const int GetNumFields = -1;
+
+        // An override of this method will be injected by the compiler into all valuetypes that cannot be compared
+        // using a simple memory comparison.
+        // This API is a bit awkward because we want to avoid burning more than one vtable slot on this.
+        // When index == GetNumFields, this method is expected to return the number of fields of this
+        // valuetype. Otherwise, it returns the offset and type handle of the index-th field on this type.
+        internal virtual int __GetFieldHelper(int index, out EETypePtr eeType)
+        {
+            // Value types that don't override this method will use the fast path that looks at bytes, not fields.
+            Debug.Assert(index == GetNumFields);
+            eeType = default;
+            return UseFastHelper;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj == null || obj.EETypePtr != this.EETypePtr)
+                return false;
+
+            int numFields = __GetFieldHelper(GetNumFields, out _);
+
+            ref byte thisRawData = ref this.GetRawData();
+            ref byte thatRawData = ref obj.GetRawData();
+
+            if (numFields == UseFastHelper)
+            {
+                // Sanity check - if there are GC references, we should not be comparing bytes
+                Debug.Assert(!this.EETypePtr.HasPointers);
+
+                // Compare the memory
+                int valueTypeSize = (int)this.EETypePtr.ValueTypeSize;
+                for (int i = 0; i < valueTypeSize; i++)
```

Comment:
Is this really the efficient way to do this? I thought a lot of work went into making Span comparison fast

Reply:
It is not available it CoreLib yet. It will be after https://github.com/dotnet/coreclr/pull/16521.